src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2021 Hunter Monroe. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hunter Monroe, Kyle Miller, Alena Gusakov
+-/
+
+import combinatorics.simple_graph.basic
+
+/-!
+# Subgraphs of a simple graph
+
+This file formalizes a subgraph of a simple graph as a structure and proves a few simple lemmas.
+-/
+
+universe u
+variables {V : Type u} (G : simple_graph V)
+open simple_graph
+
+/--
+A subgraph of a graph is a subset of vertices and a subset of edges such that each endpoint of
+an edge is contained in the vertex set. Subgraphs implement the `simple_graph` class.
+-/
+@[ext]
+structure subgraph :=
+(verts : set V)
+(adj' : V → V → Prop)
+(adj_sub : ∀ ⦃v w : V⦄, adj' v w → G.adj v w)
+(edge_vert : ∀ ⦃v w : V⦄, adj' v w → v ∈ verts)
+(sym' : symmetric adj')
+
+/--
+The Prop that states that `G'` is isomorphic to a subgraph of `G`.
+-/
+def is_subgraph (G' : simple_graph V) : Prop := ∀ ⦃v w : V⦄, G'.adj v w → G.adj v w
+
+namespace subgraph
+
+variable {G}
+
+/--
+The edges of `G'` consist of a subset of edges of `G`.
+-/
+def edge_set' (G' : subgraph G) : set (sym2 V) := sym2.from_rel G'.sym'
+
+@[simp]
+lemma mem_edge_set' {G' : subgraph G} {v w : V} : ⟦(v, w)⟧ ∈ edge_set' G' ↔ G'.adj' v w :=
+by refl
+
+lemma edge_sub (G' : subgraph G) : G'.edge_set' ⊆ G.edge_set :=
+begin
+  intro e,
+  refine quotient.rec_on_subsingleton e (λ e, _),
+  cases e with v w,
+  simp only [mem_edge_set', mem_edge_set],
+  apply G'.adj_sub,
+end
+
+lemma has_verts (G' : subgraph G) : ∀ ⦃e : sym2 V⦄ ⦃v : V⦄, e ∈ G'.edge_set' → v ∈ e →
+  v ∈ G'.verts :=
+begin
+  intro e,
+  refine quotient.rec_on_subsingleton e (λ e, _),
+  cases e with v w, intros u he hu,
+  simp only [mem_edge_set'] at he,
+  cases sym2.mem_iff.mp hu; subst u,
+  exact G'.edge_vert he,
+  exact G'.edge_vert (G'.sym' he),
+end
+
+lemma adj_symm' (G' : subgraph G) ⦃v w : V⦄ : G'.adj' v w ↔ G'.adj' w v :=
+by { split; apply G'.sym' }
+
+/--
+Function lifting `G' : subgraph G` to `G' : simple_graph G.V`
+-/
+def to_simple_graph {G : simple_graph V} (G' : subgraph G) : simple_graph G'.verts :=
+{ adj := λ v w, G'.adj' ↑v ↑w,
+  sym := λ v w h, G'.sym' h,
+  loopless := λ ⟨v, _⟩ h, loopless G v (G'.adj_sub h) }
+
+/--
+Coercion from `G' : subgraph G` to `G' : simple_graph G.V`
+-/
+def coe {V : Type*} {G : simple_graph V} (G' : subgraph G) : simple_graph V :=
+{ adj := G'.adj',
+  sym := G'.sym',
+  loopless := λ v h, loopless G v (G'.adj_sub h) }
+
+/--
+The subgraph type is inhabited by a graph with no vertices.
+-/
+instance : inhabited (subgraph G) := { default :=
+{ verts := ∅,
+  adj' := λ v w, false,
```

Comment:
```suggestion
  adj' := ⊥,
```
Also, I think usually instead of writing the whole `{ default :=` for the `inhabited` constructor, mathlib usually uses angle brackets.

Reply:
Sorry I could not make this work with ` adj' := ⊥` or angle brackets, could you please provide code?

Reply:
`⊥` seems like it should work immediately, but what's happening is that `⊥ v w` simplifies to `∅ w`, which is misusing the set interface.  (Maybe @eric-wieser has insight?  I'd thought that relations should have the reasonable bounded lattice instance, and indeed they do, but somehow simplification goes awry.)

This can be fixed with `change`:
```lean
instance : inhabited (subgraph G) :=
⟨{ verts := ∅,
   adj' := ⊥,
   adj_sub := λ v w, by { change false → _, simp, },
   edge_vert := λ v w, by { change false → _, simp, },
   sym' := λ v w, by { change false → _, simp, } }⟩
```

Using the original `adj'`:
```lean
instance : inhabited (subgraph G) :=
⟨{ verts := ∅,
   adj' := λ v w, false,
   adj_sub := λ v w, by simp,
   edge_vert := λ v w, by simp,
   sym' := λ v w, by simp }⟩
```